tests/robustness/multiclient_test/framework/harness.go

```diff
@@ -0,0 +1,241 @@
+	dataRepoPath := path.Join(*repoPathPrefix, dataSubPath)
+	metadataRepoPath := path.Join(*repoPathPrefix, metadataSubPath)
+
+	th := &TestHarness{}
+	th.init(ctx, dataRepoPath, metadataRepoPath)
+
+	return th
+}
+
+// TestHarness provides a Kopia robustness.Engine.
+type TestHarness struct {
+	dataRepoPath string
+	metaRepoPath string
+
+	baseDirPath string
+	fileWriter  *MultiClientFileWriter
+	snapshotter *MultiClientSnapshotter
+	persister   *snapmeta.KopiaPersister
+	engine      *engine.Engine
+
+	skipTest bool
+}
+
+func (th *TestHarness) init(ctx context.Context, dataRepoPath, metaRepoPath string) {
+	th.dataRepoPath = dataRepoPath
+	th.metaRepoPath = metaRepoPath
+
+	// Override ENGINE_MODE env variable. Multiclient tests can only run in SERVER mode.
+	log.Printf("Setting %s to %s\n", snapmeta.EngineModeEnvKey, snapmeta.EngineModeServer)
+
+	err := os.Setenv(snapmeta.EngineModeEnvKey, snapmeta.EngineModeServer)
+	if err != nil {
+		log.Printf("Error setting ENGINE_MODE to server: %s", err.Error())
+		os.Exit(1)
+	}
+
+	// the initialization state machine is linear and bails out on first failure
+	if th.makeBaseDir() && th.getFileWriter() && th.getSnapshotter() &&
+		th.getPersister() && th.getEngine(ctx) {
+		return // success!
+	}
+
+	err = th.Cleanup(ctx)
+	if err != nil {
+		log.Printf("Error cleaning up the engine: %s\n", err.Error())
+		os.Exit(2)
+	}
+
+	if th.skipTest {
+		os.Exit(0)
+	}
+
+	os.Exit(1)
+}
+
+func (th *TestHarness) makeBaseDir() bool {
+	baseDir, err := ioutil.TempDir("", "engine-data-")
+	if err != nil {
+		log.Println("Error creating temp dir:", err)
+		return false
+	}
+
+	th.baseDirPath = baseDir
+
+	return true
+}
+
+func (th *TestHarness) getFileWriter() bool {
+	fw, err := NewMultiClientFileWriter(
+		func() (FileWriter, error) { return fiofilewriter.New() },
+	)
+	if err != nil {
+		if errors.Is(err, fio.ErrEnvNotSet) {
+			log.Println("Skipping robustness tests because FIO environment is not set")
+
+			th.skipTest = true
+		} else {
+			log.Println("Error creating fio FileWriter:", err)
+		}
+
+		return false
+	}
+
+	th.fileWriter = fw
+
+	return true
+}
+
+func (th *TestHarness) getSnapshotter() bool {
+	server, err := snapmeta.NewSnapshotter(th.baseDirPath)
+	if err != nil {
+		if errors.Is(err, kopiarunner.ErrExeVariableNotSet) {
+			log.Println("Skipping robustness tests because KOPIA_EXE is not set")
+
+			th.skipTest = true
+		} else {
+			log.Println("Error creating kopia Snapshotter:", err)
+		}
+
+		return false
+	}
+
+	newClientFn := func(baseDirPath string) (ClientSnapshotter, error) {
+		return snapmeta.NewSnapshotter(th.baseDirPath)
+	}
+
+	s, err := NewMultiClientSnapshotter(th.baseDirPath, newClientFn, server)
+	if err != nil {
+		log.Println("Error creating multiclient kopia Snapshotter:", err)
+		return false
+	}
+
+	th.snapshotter = s
+
+	if err = s.ConnectOrCreateRepo(th.dataRepoPath); err != nil {
+		log.Println("Error initializing kopia Snapshotter:", err)
+		return false
+	}
```

Comment:
This is a nitpick, but this construction sequence is not modular nor elegant:

- you create a real Snapshotter ("server")
- you create a MultiClientSnapshotter, passing it the real Snapshotter
- you call ConnectOrCreateRepo on the MultiClientSnapshotter which internally invokes the same on the real Snapshotter followed by some policy call

Why expose the real Snapshotter here? Why not make the MultiClientSnapshotter constructor create the real Snapshotter ("server") internally?

Reply:
Alternatively, provide MultiClientSnapshotter a fully initialized real Snapshotter.